### DISCUSSION TITLE
Best way to correct dependency parsing mistakes

### DISCUSSION THREAD

**Discussion by jama1017** [explosion/spaCy#13723] [Help: Best practices]
Hi,

I am using SpaCy's built-in `en_core_web_trf` to parse out texts that describe a motion instruction, for example "Move the black rock upward by 50 miles" and "Move the black rock north by 50 miles" 

In the first example, spaCy would correctly label the preposition "by" as "prep" depending on the verb "Move". However, in the second example, since "upward" has been changed to the noun "north", spaCy would label "by" as "prep" depending on the noun "north", instead of "move" like it previously did.

There are many other examples like this in my data, and I wonder how I can correct these parsing mistakes. Is it through training from scratch, or fine-tuning, or pseudo rehearsal? I tried to search for tutorials on how to fine-tune the DependencyParser but didn't find any up-to-date ones. Any ideas and pointers would be helpful. Thanks so much!

**Comment by jama1017** [comment]
Just to add onto this - I wonder if it's possible to train a dependency parser from scratch with OntoNotes 5.0 (I believe this is what the out of the box `en_core_web_trf` was being trained on) but also with some custom data? Would love some guidance on this if people have done this before!!

**Reply by nrodnova** [reply]
You can totally train spaCy from scratch if you have access to OntoNotes. 
Fine-tuning worked just fine for me. I didn't even bother with transformers, just used the large model.
The steps I took were:
1. Select a representative set of texts (i.e. with mistakes, but keep the "good stuff" there as well)
2. Split into sentences (to make it manageable)
3. Load into Prodigy with the dependencies the model produced (one sentence per doc)
4. Correct the dependencies
5. Assemble back annotated sentences into larger docs
6. Fine-tune the model
Worked pretty well for me.I could see the improvements after just a couple of hundreds of examples. Of course, the more the merrier.

**Comment by ivan-kleshnin** [comment]
I second this question. The largest amount of mistakes Spacy does, in my experience, are related to sentence dependencies.
For example, in my case, "co-founder" is constantly mislabeled in a way that `founder` is subservient to `co`, which is clearly should be vice-versa, or sometimes it mislables them such as both words are unrelated. `sm`, `md`, `lg` models – all share this mistake.

I would like more guidance on how:
- add manual exceptions or corrections for quick fixes
- more guidance on model retraining for sentence parts. The only doable approach I see at the moment is to work from existing examples, e.g I find a correct sentence closest to some of my incorrect cases and get clues from there. Because manually setting ALL the data seems an extremely hard tasks that requires a working linguistics experience.

**Reply by nrodnova** [reply]
The thing with "co-founder"-like issues is that in OntoNotes (on which spaCy's parser was trained on), keeps it as one word, while spaCy's tokenizer splits it into 3 tokens. So there's not really any training data for it to behave the way you expect.
From what I recall, there are only about 2K unique words containing '-' that are a single token in OntoNotes. So if you are determined (and consistent) in how to attach a particle to the main word, you can do it programmatically (if you have OntoNotes) and re-train the model.
Also, from my limited experiments, creating a small dataset focused on the things you want to correct, annotating it in Prodigy, and and then just fine-tuning the existing parser, works pretty well.

**Reply by ivan-kleshnin** [reply]
Thanks, Natalia. I don't have an access to Prodigy yet. OntoNotes is not public either. I'm quite determined to retrain the model, the issue is that, it seems, there's no original data (on which Spacy was trained on) in public. Only some scattered fragments. https://spacy.io/usage/projects have examples with UD dependencies which I don't like and currently plan to avoid.

- Is there any link to download the original data (for `tagger` and `parser` pipes)?
- if not, can you suggest an alternative resource with open, non UD-marked, treebanks? Preferably in JSON or CONLL-U formats.

I'm checking Kaggle and it's all trash so far (untagged data, viruses, etc). I wonder how we're suggested to "mix the original data to avoid forgetting" but there is no "original data" anywhere to find 🤷‍♂️  I know I can feed the output of the trained model back to retrain another one... but I'd rather pick the original, correct data instead of potentially (and practically) incorrect interpretation of it.

https://huggingface.co is also full of anything but what I need.
